### ISSUE TITLE
Broadcaster.reduce() doesn't complete if stream has mapper that returns (enough) null objects

### ISSUE THREAD

**Issue by make** [reactor/reactor#573]
Having a mapper that sets null objects in stream breaks the Broadcaster even if null objects are filtered in next step.

dependencies:
```groovy
	compile 'io.projectreactor:reactor-core:2.0.7.RELEASE'
	compile 'io.projectreactor:reactor-stream:2.0.7.RELEASE'
```

This doesn't get to consume()
```groovy
import reactor.Environment
import reactor.core.Dispatcher
import reactor.fn.Supplier
import reactor.rx.broadcast.Broadcaster

	def test() {
		def errorHandler = { err ->
			log.error("Error in stream", err)
			throw err
		}
		
		Supplier<Dispatcher> supplier = Environment.newCachedDispatchers(1, 'reactor-test')

		Broadcaster source = Broadcaster.create(supplier.get());
		source
			.map { Map m ->
				m.integer % 3 == 0 ? null : m
			}
			.filter { Map m ->
				m?.integer < 5
			}
			.reduce([], { a, b ->
				log.info "Reducing $a, $b"
				a + b
			})
			.consume({ l ->
				println("\n\nResults: $l")
			}, errorHandler)
	
		int integers = 1000000
		for(i in 1..integers) {
			if(i % 100000 == 0)
				log.info i
			source.onNext([integer:i])
		}
		log.info "Passed $integers integers into stream"
		source.onComplete()
	}
```

output
```
2015-11-25 14:10:20,553 [reactor-test-208] INFO .a.c.d.m.ReactorTestController - Reducing [], [integer:1]
2015-11-25 14:10:20,554 [reactor-test-208] INFO .a.c.d.m.ReactorTestController - Reducing [[integer:1]], [integer:2]
2015-11-25 14:10:20,554 [reactor-test-208] INFO .a.c.d.m.ReactorTestController - Reducing [[integer:1], [integer:2]], [integer:4]
2015-11-25 14:10:20,647 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 100000
2015-11-25 14:10:20,694 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 200000
2015-11-25 14:10:20,742 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 300000
2015-11-25 14:10:20,788 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 400000
2015-11-25 14:10:20,835 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 500000
2015-11-25 14:10:20,884 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 600000
2015-11-25 14:10:20,930 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 700000
2015-11-25 14:10:20,977 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 800000
2015-11-25 14:10:21,024 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 900000
2015-11-25 14:10:21,071 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - 1000000
2015-11-25 14:10:21,071 [http-nio-8080-exec-6] INFO .a.c.d.m.ReactorTestController - Passed 1000000 integers into stream
```

----

But the code below works as expected
```groovy
import reactor.Environment
import reactor.core.Dispatcher
import reactor.fn.Supplier
import reactor.rx.broadcast.Broadcaster

	def test() {
		def errorHandler = { err ->
			log.error("Error in stream", err)
			throw err
		}
		
		Supplier<Dispatcher> supplier = Environment.newCachedDispatchers(1, 'reactor-test')

		Broadcaster source = Broadcaster.create(supplier.get());
		source
			.filter { Map m ->
				m?.integer < 5
			}
			.reduce([], { a, b ->
				log.info "Reducing $a, $b"
				a + b
			})
			.consume({ l ->
				println("\n\nResults: $l")
			}, errorHandler)
	
		int integers = 1000000
		for(i in 1..integers) {
			if(i % 100000 == 0)
				log.info i
			source.onNext([integer:i])
		}
		log.info "Passed $integers integers into stream"
		source.onComplete()
	}
```

output from latter:
```
2015-11-25 13:59:03,474 [reactor-test-210] INFO .a.c.d.m.ReactorTestController - Reducing [], [integer:1]
2015-11-25 13:59:03,474 [reactor-test-210] INFO .a.c.d.m.ReactorTestController - Reducing [[integer:1]], [integer:2]
2015-11-25 13:59:03,474 [reactor-test-210] INFO .a.c.d.m.ReactorTestController - Reducing [[integer:1], [integer:2]], [integer:3]
2015-11-25 13:59:03,474 [reactor-test-210] INFO .a.c.d.m.ReactorTestController - Reducing [[integer:1], [integer:2], [integer:3]], [integer:4]
2015-11-25 13:59:04,297 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 100000
2015-11-25 13:59:05,119 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 200000
2015-11-25 13:59:05,934 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 300000
2015-11-25 13:59:06,775 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 400000
2015-11-25 13:59:07,598 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 500000
2015-11-25 13:59:08,413 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 600000
2015-11-25 13:59:09,248 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 700000
2015-11-25 13:59:10,068 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 800000
2015-11-25 13:59:10,881 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 900000
2015-11-25 13:59:11,711 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - 1000000
2015-11-25 13:59:11,711 [http-nio-8080-exec-3] INFO .a.c.d.m.ReactorTestController - Passed 1000000 integers into stream


Results: [[integer:1], [integer:2], [integer:3], [integer:4]]
```

**Comment by make** [comment]
Notice also that the failing example took only 518 ms to pass 1M objects into the stream while the working example took over 8s to do the same.

**Comment by smaldini** [comment]
Problem is that null is a forbidden value within the Reactive Streams Standard I wonder if you face the same with Master (2.1.0.BUILD-SNAPSHOT). Just replace 
```
  Supplier<Dispatcher> supplier = Environment.newCachedDispatchers(1, 'reactor-test')
```

by 
```
  Supplier<Processor> supplier = Processors.asyncGroup('reactor-test', 128, 1)
```

**Comment by smaldini** [comment]
Fixed in 2.5.0 (M1 to come, snapshot available)
http://next.projectreactor.io/stream/docs/api/reactor/rx/Stream.html.
